x/msgfee/decorator.go

```diff
@@ -0,0 +1,78 @@
+package msgfee
+
+import (
+	"github.com/iov-one/weave"
+	"github.com/iov-one/weave/coin"
+	"github.com/iov-one/weave/errors"
+)
+
+// FeeDecorator implements a decorator that for each processed transaction
+// attach an additional fee to the result. Each fee is declarated per
+// transaction type. If fee is not set (zero value) then this decorator does
+// not increase the required fee value.
+// Additional fee is attached to only those transaction results that represent
+// a success.
+type FeeDecorator struct {
+	bucket *MsgFeeBucket
+}
+
+var _ weave.Decorator = (*FeeDecorator)(nil)
+
+// NewFeeDecorator returns a decorator that is upading the cost of processing
+// each message according to the fee configured per each message type.
+func NewFeeDecorator() *FeeDecorator {
+	return &FeeDecorator{
+		bucket: NewMsgFeeBucket(),
+	}
+}
+
+func (d *FeeDecorator) Check(ctx weave.Context, store weave.KVStore, tx weave.Tx, next weave.Checker) (weave.CheckResult, error) {
+	res, err := next.Check(ctx, store, tx)
+	if err != nil {
+		return res, err
+	}
+
+	switch fee, err := txFee(d.bucket, store, tx); {
```

Comment:
why this switch/case instead of two if statements?
maybe a line or two shorter (`}` lines), but seems more non-standard and harder to read.

Can you just make it `if err != nil { ...} if !coin.IsEmpty(fee) {...}`?

Reply:
Updated in https://github.com/iov-one/weave/pull/376/commits/ead18fc79123fe870bfb33db233509d6889ebaed